### PR TITLE
[WIP] Improve typing

### PR DESCRIPTION
**Pull request by abihf**
I add generic type to define field names that are being used by the form.

Usage:
```typescript

type FormDataType = {
  name: string
  email: string
}

...

{ handleSubmit, errors, setValue } = useForm<FormDataType>()

const onSubmit = handleSubmit(({ name, email }) => {}) // name and email will have correct type

setValue("name", 123) // error: 123 is not string
errors.something.message // error: property something does not exist
```

### CODE REVIEW

File: src/types.ts [hunk 1]
```diff
@@ -57,31 +59,31 @@ export interface Error {
...
   register: RegisterFunction;
-  handleSubmit: (func: OnSubmit) => any;
-  errors: ErrorMessages;
+  handleSubmit: (func: OnSubmit<Data>) => any;
+  errors: ErrorMessages<Data>;
   watch: (
-    filedNames?: string | string[] | undefined,
-    defaultValue?: string | string[] | undefined,
+    filedNames?: keyof Data | Array<keyof Data> | undefined,
+    defaultValue?: keyof Data | Array<keyof Data> | undefined,
```
Comment (abihf) [inline src/types.ts]: TODO: fix defaultValue type

**Comment by bluebill1049** [comment]
OMG! this is so good. I am learning by looking at your PR. Thank you so much.

**Comment by bluebill1049** [comment]
thanks very much, @abihf :) can you please fix some of the errors from eslint? we can leave those warnings tho :)

`yarn lint` 
 
Screen Shot 2019-05-26 at 6 48 53 pm

cheers
bill

**Comment by abihf** [comment]
Sure.
Will continue my PR tomorrow.

File: src/index.ts [hunk 2]
```diff
@@ -319,7 +321,7 @@ export default function useForm(
   return {
     register,
     handleSubmit,
-    watch,
+    watch: watch as WatchFunction<Data>,
```
Comment (bluebill1049) [inline src/index.ts]: curious if we can move this to line: 175

**Comment by bluebill1049** [comment]
love those generic types. great work mate 👍

File: src/types.ts [hunk 3]
```diff
@@ -57,31 +59,32 @@ export interface Error {
...
+  errors: ErrorMessages<Data>;
+  values: Data;
 }
 
 export type VoidFunction = () => void;
 
 export type RegisterFunction = (refOrValidateRule: RegisterInput | Ref, validateRule?: RegisterInput) => any;
 
-export interface UseFormFunctions {
+export type WatchFunction<Data extends DataType> = ((name?: undefined) => void)
```
Comment (bluebill1049) [inline src/types.ts]: if the `name` is undefined, it should return all data.

`watch()` => {test: 1, test: 2, test:3}
`watch('test')` => 1
`watch(['test', 'test2']})` => {test: 1, test: 2}

**Comment by bluebill1049** [comment]
hey @abihf 

If you are busy with things, let me know I am happy to merge it in and finish the rest of my comments in my own time. again thanks for your great work! #typescriptMaster :)

cheers
bill

**Comment by abihf** [comment]
Hi @bluebill1049 

Sorry, I just got time to fix the eslint error. Is that enaugh?

**Comment by abihf** [comment]
Btw, I updated `package-lock.json`, I thought you use npm for installing dependencies. Do I need to revert it back?

**Comment by bluebill1049** [comment]
awesome!! merging I will take care of the rest. I will add you to the contributors' list if you don't mind.

thank you very much.

**Pull request merged**